lib/cn/ltok.c

```diff
@@ -0,0 +1,203 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#define MTF_MOCK_IMPL_ltok
+
+#include "ltok.h"
+
+#if HSE_MOCKING
+#include "ltok_ut_impl.i"
+#endif
+
+void
+ltok_init(struct ltok *ltok)
+{
+    memset(ltok, 0, sizeof(*ltok));
+    spin_lock_init(&ltok->ltok_lock);
+}
+
+static inline
+unsigned
+shared_users(struct ltok *ltok)
+{
+    return ltok->ltok_active_spills + ltok->ltok_active_kcomps + ltok->ltok_active_kvcomps;
+}
+
+static inline
+bool
+is_idle(struct ltok *ltok)
+{
+    return shared_users(ltok) == 0 && !ltok->ltok_active_join && !ltok->ltok_active_split;
+}
+
+static inline
+bool
+allow_shared(struct ltok *ltok, uint8_t *counter)
+{
+    if (ltok->ltok_active_join || ltok->ltok_active_split || ltok->ltok_reserved)
+        return false;
+
+    if (*counter == UINT8_MAX)
+        return false;
+
+    *counter += 1;
+    return true;
+}
+
+static inline
+bool
+allow_exclusive(struct ltok *ltok, bool *granted)
+{
+    if (is_idle(ltok) && !ltok->ltok_reserved) {
+        *granted = true;
+        return true;
+    }
+
+    return false;
+}
+
+bool
+ltok_get(struct ltok *ltok, enum ltok_token token)
```

Comment:
Would be good to add invariants that ltok isn't NULL.

Given that ltok is private to cN, I see no reason to bubble an merr_t